networkx/algorithms/approximation/traveling_salesman.py

```diff
@@ -372,7 +371,290 @@ def asadpour_tsp(G, weight="weight"):
+            # We have to pick the root node of the arborescence for the out
+            # edge of the first vertex as that is the only node without an
+            # edge directed into it.
+            for N in arb:
+                if arb.in_degree(N) == 0:
+                    # root found
+                    arb.add_edge(n, N, weight=G[n][N][weight])
+                    arb_weight += G[n][N][weight]
+                    break
+            # We can pick the minimum weight in-edge for the vertex with a
+            # cycle
+            min_in_edge_weight = math.inf
+            min_in_edge = None
+            for u, v, d in G.in_edges(n, data=True):
+                edge_weight = d[weight]
+                if u == N:
+                    continue
+                if edge_weight < min_in_edge_weight:
+                    min_in_edge_weight = edge_weight
+                    min_in_edge = (u, v)
+            arb.add_edge(min_in_edge[0], min_in_edge[1], weight=min_in_edge_weight)
+            arb_weight += min_in_edge_weight
+            # Check to see the weight of the arborescence, if it is a new
+            # minimum, clear all of the old potential minimum
+            # 1-arborescences and add this is the only one. If its weight is
+            # above the known minimum, do not add it.
+            if arb_weight < minimum_1_arborescence_weight:
+                minimum_1_arborescences.clear()
+                minimum_1_arborescence_weight = arb_weight
+            # We have a 1-arborescence, add it to the set
+            if arb_weight == minimum_1_arborescence_weight:
+                minimum_1_arborescences.add(arb)
+
+        return minimum_1_arborescences
+
+    def direction_of_ascent():
+        """
+        Find the direction of ascent at point pi.
+
+        See [1]_ for more information.
+
+        Returns
+        -------
+        dict
+            A mapping from the nodes of the graph which represents the direction
+            of ascent.
+
+        References
+        ----------
+        .. [1] M. Held, R. M. Karp, The traveling-salesman problem and minimum
+           spanning trees, Operations Research, 1970-11-01, Vol. 18 (6),
+           pp.1138-1162
+        """
+        # 1. Set d equal to the zero n-vector.
+        d = {}
+        for n in G:
+            d[n] = 0
+        del n
+        # 2. Find a 1-Aborescence T^k such that k is in K(pi, d).
+        minimum_1_arborescences = k_pi()
+        while True:
+            # Reduce K(pi) to K(pi, d)
+            # Find the arborescence in K(pi) which increases the lest in
+            # direction d
+            min_k_d_weight = math.inf
+            min_k_d = None
+            for arborescence in minimum_1_arborescences:
+                weighted_cost = 0
+                v_k_0 = 0
+                for n in G:
+                    weighted_cost += d[n] * (arborescence.degree(n) - 2)
+                    v_k_0 += 1 if arborescence.degree(n) - 2 == 0 else 0
```

Comment:
Again, faster to iterate over `arborescence.degree`.  There's an apparent discrepancy if the nodes of `arborescence` is a proper subset of the nodes of `G`, but if that was the case, this code would fail anyway.